Riot/Modules/Room/NotificationSettings/RoomNotificationSettingsViewModel.swift

```diff
@@ -0,0 +1,89 @@
+// File created from ScreenTemplate
+// $ createScreen.sh Room/NotificationSettings RoomNotificationSettings
+/*
+ Copyright 2020 New Vector Ltd
+ 
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ 
+ http://www.apache.org/licenses/LICENSE-2.0
+ 
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+import Foundation
+
+final class RoomNotificationSettingsViewModel: RoomNotificationSettingsViewModelType {
+    
+    // MARK: - Properties
+    
+    // MARK: Private
+    
+    private let roomNotificationService: RoomNotificationSettingsServiceType
+    private var state: RoomNotificationSettingsViewState {
+        willSet {
+            update(viewState: newValue)
+        }
+    }
+    // MARK: Public
+
+    weak var viewDelegate: RoomNotificationSettingsViewModelViewDelegate?
+    
+    weak var coordinatorDelegate: RoomNotificationSettingsViewModelCoordinatorDelegate?
+    
+    // MARK: - Setup
+    
+    init(roomNotificationService: RoomNotificationSettingsServiceType, roomEncrypted: Bool, avatarViewData: AvatarViewDataProtocol?) {
+        self.roomNotificationService = roomNotificationService
+        
+        let notificationState = Self.mapNotificationStateOnRead(encrypted: roomEncrypted, state: roomNotificationService.notificationState)
+        self.state = RoomNotificationSettingsViewState(roomEncrypted: roomEncrypted, saving: false, notificationState: notificationState, avatarData: avatarViewData)
+        self.roomNotificationService.observeNotificationState { [weak self] state in
+            guard let self = self else { return }
+            
+            self.state.notificationState = Self.mapNotificationStateOnRead(encrypted: roomEncrypted, state: state)
+        }
```

Comment:
I think we should call `didUpdateViewState` to notify the view about the new state.

Reply:
It gets called by the [willSet](https://github.com/vector-im/element-ios/pull/4458/files/15917907b7143da3881299bf93b9b2c0bde9d65f#diff-20eb2dc7b5e30713d0afb03f1ebdf4de3d712d785d55bab260e39c5cfeb7dec4R29). 'self.state' being a value type, mutating a part of it changes the structs value as a whole ,causing 'willSet' to trigger. It's a patten I have used for managing state changes and triggering the ui update without too much boilerplate. If it's something we think is a bit too implicit we could also use something a bit more explicit like reducer functions ala elm/redux, which would usually be my preference if the ViewModel logic gets larger/more complex.

Reply:
We want to move to reducer approach to have more deterministic states.
I definitely prefer your `willSet` implementation. Our template are too verbose for us at the moment.